test/stdlib/Kernel_test.rb

```diff
@@ -11,73 +11,208 @@ def test_Array
+    assert_send_type "(String) -> Complex",
+                     Kernel, :Complex, '1'
+    assert_send_type "(String, exception: true) -> Complex",
+                     Kernel, :Complex, '1', exception: true
+
+    # (untyped real_or_both, exception: bool) -> Complex?
+    with_untyped.and 'oops' do |real_untype|
+      assert_send_type '(untyped, exception: bool) -> Complex?',
+                       Kernel, :Complex, real_untype, exception: false
+    end
+
+    with '1', 1, 1r, 1.0, (1+0i), numeric do |real|
+      with '2', 2, 2r, 2.0, (2+0i), numeric do |imag|
+        # (Numeric | String real, Numeric | String imag, ?exception: true) -> Complex
+        assert_send_type "(Numeric | String, Numeric | String) -> Complex",
+                         Kernel, :Complex, real, imag
+        assert_send_type "(Numeric | String, Numeric | String, exception: true) -> Complex",
+                         Kernel, :Complex, real, imag, exception: true
+
+        # Complex has an awkward edgecase where `exception: false` will unconditionally return `nil`
+        # if the imaginary argument is not one of the builtin `Numeric`s. Oddly enough, it's not for
+        # the `real` one...
+        case imag
+        when Integer, Float, Rational, Complex
+          # (Numeric | String real, Integer | Float | Rational | Complex imag, exception: bool) -> Complex
+          assert_send_type "(Numeric | String, Integer | Float | Rational | Complex, exception: bool) -> Complex",
+                           Kernel, :Complex, real, imag, exception: false
+        end
+      end
+
+      # (Numeric | String real, untyped, exception: bool) -> Complex?
+      with_untyped.and 'oops', numeric do |imag|
+        next if [Integer, Float, Rational, Complex].any? { _1 === imag }
+        assert_send_type "(Numeric | String, untyped, exception: bool) -> nil",
+                         Kernel, :Complex, real, imag, exception: false
+      end
+    end
   end
 
+
   def test_Float
-    with_float 1.0 do |float|
-      assert_send_type "(::float) -> Float",
-                       Kernel, :Float, float
-      assert_send_type "(::float, exception: true) -> Float",
-                       Kernel, :Float, float, exception: true
-      assert_send_type "(::float, exception: bool) -> Float?",
-                       Kernel, :Float, float, exception: false
+    with 1, 1.0, ToF.new(1.0), '1e3' do |float_like|
+      assert_send_type "(_ToF) -> Float",
+                       Kernel, :Float, float_like
+      assert_send_type "(_ToF, exception: true) -> Float",
+                       Kernel, :Float, float_like, exception: true
+      assert_send_type "(_ToF, exception: bool) -> Float",
+                       Kernel, :Float, float_like, exception: false
     end
 
-    assert_send_type "(untyped, ?exception: bool) -> Float?",
-                     Kernel, :Float, :hello, exception: false
+    with_untyped do |untyped|
+      next if defined? untyped.to_f
```

Comment:
I can't determine if the subsequent tests are actually being executed.

To determine this, it would be necessary to rely on the implementation of `with_untyped`.

A BasicObject, or some other object that is guaranteed not to have a `to_f` method, should be explicitly specified.